examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1089 @@
+      ordered_cells.push_back(cell->index());
+
+    // shuffle the elements
+    std::mt19937 random_number_generator;
+    for (unsigned int i = 1; i < n_cells; ++i)
+      {
+        // get a random number between 0 and i (inclusive)
+        const unsigned int j =
+          std::uniform_int_distribution<>(0, i)(random_number_generator);
+
+        // if possible, swap the elements
+        if (i != j)
+          std::swap(ordered_cells[i], ordered_cells[j]);
+      }
+
+    return ordered_cells;
+  }
+
+
+
+  template <int dim>
+  class RightHandSide : public Function<dim>
+  {
+  public:
+    RightHandSide()
+      : Function<dim>()
+    {}
+
+    virtual double value(const Point<dim> & p,
+                         const unsigned int component = 0) const;
+
+    virtual void value_list(const std::vector<Point<dim>> &points,
+                            std::vector<double> &          values,
+                            const unsigned int             component = 0) const;
+  };
+
+  template <int dim>
+  double RightHandSide<dim>::value(const Point<dim> & p,
+                                   const unsigned int component) const
+  {
+    Assert(component == 0, ExcIndexRange(component, 0, 1));
+    (void)component;
+    (void)p;
+
+    return 0.0;
+  }
+
+
+
+  template <int dim>
+  void RightHandSide<dim>::value_list(const std::vector<Point<dim>> &points,
+                                      std::vector<double> &          values,
+                                      const unsigned int component) const
+  {
+    Assert(values.size() == points.size(),
+           ExcDimensionMismatch(values.size(), points.size()));
+
+    for (unsigned int i = 0; i < points.size(); ++i)
+      values[i] = RightHandSide<dim>::value(points[i], component);
+  }
+
+
+
+  template <int dim>
+  class BoundaryValues : public Function<dim>
+  {
+  public:
+    BoundaryValues()
+      : Function<dim>()
+    {}
+
+    virtual double value(const Point<dim> & p,
+                         const unsigned int component = 0) const;
+
+    virtual void value_list(const std::vector<Point<dim>> &points,
+                            std::vector<double> &          values,
+                            const unsigned int             component = 0) const;
+  };
+
+
+  template <int dim>
+  double BoundaryValues<dim>::value(const Point<dim> & p,
+                                    const unsigned int component) const
+  {
+    Assert(component == 0, ExcIndexRange(component, 0, 1));
+    (void)component;
+
+    if (std::fabs(p[0] * p[0] + p[1] * p[1] - 0.3 * 0.3) <
+          1e-8                                        // around cylinder
+        || std::fabs(p[0] + 1) < 1e-8                 // x == -1
+        || std::fabs(p[1] - 1) < 1e-8                 // y == 1
+        || (std::fabs(p[1] + 1) < 1e-8 && p[0] < 0.5) // y == -1, x <= 0.5
+    )
+      {
+        return 0.0;
+      }
+    else if (std::fabs(p[0] - 1) < 1e-8                     // x = 1
```

Comment:
strictly, speaking, you only need this check here, as the rest of the values is 0.0, right?

Reply:
Good point! I changed an re-ran to confirm we were solving the same problem.